python/lsst/pipe/tasks/dcrModel.py

```diff
@@ -0,0 +1,539 @@
+        regularizeSigma : `float`
+            Threshold to exclude noise-like pixels from regularization.
+        clampFrequency : `float`
+            Maximum relative change of the model allowed between subfilters.
+        convergenceMaskPlanes : `list` of `str`, or `str`, optional
+            Mask planes to use to calculate convergence.
+            Default value is set in ``calculateNoiseCutoff`` if not supplied.
+        """
+        templateImage = self.getReferenceImage(bbox)
+        excess = np.zeros_like(templateImage)
+        for model in self:
+            noiseCutoff = self.calculateNoiseCutoff(model, statsCtrl, regularizeSigma,
+                                                    mask=mask[bbox],
+                                                    convergenceMaskPlanes=convergenceMaskPlanes)
+            modelVals = model.image.array
+            highPixels = (modelVals > (templateImage*clampFrequency + noiseCutoff))
+            excess[highPixels] += modelVals[highPixels] - templateImage[highPixels]*clampFrequency
+            modelVals[highPixels] = templateImage[highPixels]*clampFrequency
+            lowPixels = (modelVals < templateImage/clampFrequency - noiseCutoff)
+            excess[lowPixels] += modelVals[lowPixels] - templateImage[lowPixels]/clampFrequency
+            modelVals[lowPixels] = templateImage[lowPixels]/clampFrequency
+        excess /= len(self)
+        for model in self:
+            model.image.array += excess
+
+    def calculateNoiseCutoff(self, maskedImage, statsCtrl, regularizeSigma, mask=None,
+                             convergenceMaskPlanes=None):
+        """Helper function to calculate the background noise level of an image.
+
+        Parameters
+        ----------
+        maskedImage : `lsst.afw.image.MaskedImage`
+            The input image to evaluate the background noise properties.
+        statsCtrl : `lsst.afw.math.StatisticsControl`
+            Statistics control object for coadd
+        regularizeSigma : `float`
+            Threshold to exclude noise-like pixels from regularization.
+        mask : `lsst.afw.image.Mask`, Optional
+            Optional alternate mask
+        convergenceMaskPlanes : `list` of `str`, or `str`, optional
+            Mask planes to use to calculate convergence.
+            The "DETECTED" mask plane is used by default if not supplied.
+
+        Returns
+        -------
+        float
+            The threshold value to treat pixels as noise in an image..
+        """
+        if convergenceMaskPlanes is None:
+            convergeMask = maskedImage.mask.getPlaneBitMask("DETECTED")
+        else:
+            convergeMask = maskedImage.mask.getPlaneBitMask(convergenceMaskPlanes)
+        if mask is None:
+            backgroundPixels = maskedImage.mask.array & (statsCtrl.getAndMask() | convergeMask) == 0
+        else:
+            backgroundPixels = mask.array & (statsCtrl.getAndMask() | convergeMask) == 0
```

Comment:
if mask is None:
        mask = maskedImage.mask
    backgroundPixels = mask.array & (statsCtrl.getAndMask() | convergeMask) == 0

This reduces code duplication.